core/storage/app_feedback_report/gae_models.py

```diff
@@ -0,0 +1,659 @@
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for Oppia feedback reports from both Android and web."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+
+from core.platform import models
+import feconf
+import python_utils
+import utils
+
+(base_models,) = models.Registry.import_models([models.NAMES.base_model])
+
+datastore_services = models.Registry.import_datastore_services()
+transaction_services = models.Registry.import_transaction_services()
+
+PLATFORM_CHOICE_ANDROID = 'android'
+PLATFORM_CHOICE_WEB = 'web'
+PLATFORM_CHOICES = [PLATFORM_CHOICE_ANDROID, PLATFORM_CHOICE_WEB]
+
+REPORT_INFO_TO_REDACT = [
+    'user_feedback_other_text_input', 'event_logs', 'logcat_logs']
+
+
+class AppFeedbackReportModel(base_models.BaseModel):
+    """Model for storing feedback reports sent from learners.
+
+    Instances of this model contain information about learner's device and Oppia
+    app settings, as well as information provided by the user in the feedback
+    report.
+
+    The id of each model instance is determined by concatenating the platform,
+    the timestamp of the reports (in ms since epoch, in UTC), and a hash of a
+    string representation of a random int.
+    """
+
+    # The platform (web or Android) that the report is sent from and that the
+    # feedback corresponds to.
+    platform = datastore_services.StringProperty(
+        required=True, indexed=True, choices=PLATFORM_CHOICES)
+    # The ID of the user that scrubbed this report, if it has been scrubbed.
+    scrubbed_by = datastore_services.StringProperty(
+        required=False, indexed=True)
+    # Unique ID for the ticket this report is assigned to (see
+    # AppFeedbackReportTicketModel for how this is constructed).
+    ticket_id = datastore_services.TextProperty(default=None, indexed=False)
+    # Datetime in UTC of when the report was submitted by the user on their
+    # device. This may be much earlier than the model entity's creation date if
+    # the report was locally cached for a long time on an Android device.
+    submitted_on = datastore_services.DateTimeProperty(
+        required=True, indexed=True)
+    # The type of feedback for this report; choices are not specified as
```

Comment:
choices are not specified as --> this can be an arbitrary string, since


(This communicates more clearly what is allowed, rather than what we don't do.)

Reply:
Done